src/util/ringdelaybuffer.cpp

```diff
@@ -0,0 +1,158 @@
+*
+* Any person wishing to distribute modifications to the Software is
+* requested to send the modifications to the original developer so that
+* they can be incorporated into the canonical version. It is also
+* requested that these non-binding requests be included along with the
+* license above.
+*/
+
+#include "ringdelaybuffer.h"
+
+#include "util/math.h"
+#include "util/sample.h"
+
+using ReadableSlice = mixxx::SampleBuffer::ReadableSlice;
+using WritableSlice = mixxx::SampleBuffer::WritableSlice;
+
+namespace {
+SINT copyRing(const ReadableSlice pSourceBuffer,
+        SINT sourcePos,
+        const WritableSlice pDestBuffer,
+        SINT destPos,
+        const SINT numItems) {
+    const SINT newSourcePos = sourcePos + numItems;
+    const SINT newDestPos = destPos + numItems;
+
+    VERIFY_OR_DEBUG_ASSERT(newSourcePos <= pSourceBuffer.length() ||
+            newDestPos <= pDestBuffer.length()) {
+        return 0;
+    }
+
+    // Check to see if the copy is not contiguous.
+    if (newSourcePos > pSourceBuffer.length() ||
+            newDestPos > pDestBuffer.length()) {
+        // Copy is not contiguous.
+        SINT firstDataBlockSize = math_min(pSourceBuffer.length() - sourcePos,
+                pDestBuffer.length() - destPos);
+
+        SampleUtil::copy(pDestBuffer.data(destPos),
+                pSourceBuffer.data(sourcePos),
+                firstDataBlockSize);
+
+        sourcePos = (sourcePos + firstDataBlockSize) % pSourceBuffer.length();
+        destPos = (destPos + firstDataBlockSize) % pDestBuffer.length();
+
+        // The second data part is the start of the ring buffer.
+        SampleUtil::copy(pDestBuffer.data(destPos),
+                pSourceBuffer.data(sourcePos),
+                numItems - firstDataBlockSize);
+    } else {
+        // Copy is contiguous.
+        SampleUtil::copy(pDestBuffer.data(destPos),
+                pSourceBuffer.data(sourcePos),
+                numItems);
+    }
+
+    return numItems;
+}
+
+} // anonymous namespace
+
+RingDelayBuffer::RingDelayBuffer(SINT bufferSize)
+        : m_firstInputBuffer(true),
+          m_writePos(0),
+          m_buffer(bufferSize) {
+    // Set the ring delay buffer items to 0.
+    m_buffer.fill(0);
+}
+
+SINT RingDelayBuffer::read(CSAMPLE* pBuffer, const SINT itemsToRead, const SINT delayItems) {
+    const SINT shift = itemsToRead + delayItems;
+
+    VERIFY_OR_DEBUG_ASSERT(shift <= m_buffer.size()) {
+        return 0;
+    }
+
+    SINT readPos = m_writePos - shift;
+
+    if (readPos < 0) {
+        readPos = readPos + m_buffer.size();
+    }
+
+    return copyRing(ReadableSlice(m_buffer.data(), m_buffer.size()),
+            readPos,
+            WritableSlice(pBuffer, itemsToRead),
+            0,
+            itemsToRead);
+}
+
+SINT RingDelayBuffer::write(const CSAMPLE* pBuffer, const SINT itemsToWrite) {
+    VERIFY_OR_DEBUG_ASSERT(itemsToWrite <= m_buffer.size()) {
+        return 0;
+    }
+
+    SINT copiedItems;
```

Comment:
```suggestion
    SINT copiedItems = 0;
```
avoid potentially uninitialized memory (in case on of the below branches change).

Reply:
Okay, I'm not just sure how the checks will react on it due to my IDE shows me, that _the value is never used_. I will try it.

Reply:
Such preventive initialization will hide real errors (variable not set in one branch) from the static code analysis done by the compiler.

Reply:
To be honest, I wouldn't bet on the fact that static analysis will catch this. Technically it can, in practice it might (assuming we even set the right compiler flags). IMO its better to analyze a variable early than possibly not all. If we really want to avoid this entirely, we can do a instead use an IILE and return the value instead, then I'd be confident for the compiler to catch the issue.

Reply:
> Okay, I'm not just sure how the checks will react on it due to my IDE shows me, that _the value is never used_. I will try it.

Mhmm interesting. 
Then here is the lambda solution I mentioned:
```cpp
const int numItems = [&]() {
if (m_firstInputBuffer) {
        // If the first input buffer is written, the first sample is on the index 0.
        // Based on the checking of an available number of samples, the situation,
        // that the writing will be non-contiguous cannot occur.
        // The itemsToWrite value is multiply by 2 to
        SampleUtil::copyWithRampingGain(m_buffer.data(), pBuffer, 0.0f, 1.0f, itemsToWrite);
        m_firstInputBuffer = false;
        return itemsToWrite;
    } else {
        return copyRing(ReadableSlice(pBuffer, itemsToWrite),
                0,
                WritableSlice(m_buffer.data(), m_buffer.size()),
                m_writePos,
                itemsToWrite);
    }

}();
```

Reply:
However, all CI tests passed, but based on the previous discussion, I would prefer the lambda solution. I thought about it too and it looked like it can look overcomplicated, but when I see your solution now, IMO it is quite clean.

Reply:
Just one thing, I changed the `int` to `SINT` due to it is returned from the lambda expression and the same type returns the `RingDelayBuffer::write` function. Just please a little clear up for me, what do you mean by this phrase: _"The itemsToWrite value is multiply by 2 to"_?

Reply:
I ask myself the same. I just copy pasted the comment from the source.